src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,552 @@
+                lower_y = new_lower_y;
+            }
+
+            // Update slope derivate
+            slope_derivate += getDerivate(&current_segment);
+        }
+    }
+
+    const n = current_timestamp - first_timestamp - 1;
+    // Record the current segment
+    try appendValue(current_segment.start_value, compressed_values);
+    if (n > 1) {
+        const den = n * (n + 1) * (2 * n + 1) / 6; // Denominator of Eq. 6
+        const slope: f80 = @max(
+            @min(slope_derivate / usizeToF80(den), upper_bound_line.slope),
+            lower_bound_line.slope,
+        ); // Get optimal slope that minimizes the squared error
+
+        try getIntercept(
+            slope,
+            first_timestamp,
+            uncompressed_values[first_timestamp],
+            &current_line,
+        );
+
+        const end_value: f64 = f80Tof64(evaluate(&current_line, current_timestamp - 1));
+        try appendValue(end_value, compressed_values);
+    } else {
+        try appendValue(current_segment.end_value, compressed_values);
+    }
+
+    try appendIndex(current_timestamp, compressed_values);
+}
+
+pub fn decompress(
+    compressed_values: []const u8,
+    decompressed_values: *ArrayList(f64),
+) error{OutOfMemory}!void {
+    // TODO: Check length of uncompressed_values is modulo 32.
+
+    const compressed_lines_and_index = mem.bytesAsSlice(
+        f64,
+        compressed_values,
+    );
+
+    var current_line: Line = Line{ .slope = 0, .intercept = 0 };
+    var current_segment: Segment = Segment{
+        .start_time = 0,
+        .start_value = 0,
+        .end_time = 0,
+        .end_value = 0,
+    };
+
+    var first_timestamp: usize = 0;
+    var index: usize = 0;
+    while (index < compressed_lines_and_index.len) : (index += 3) {
+        current_segment.start_time = first_timestamp;
+        current_segment.start_value = compressed_lines_and_index[index];
+        current_segment.end_value = compressed_lines_and_index[index + 1];
+        current_segment.end_time = f64ToUsize(compressed_lines_and_index[index + 2] - 1);
+
+        if (current_segment.start_time < current_segment.end_time) {
+            try getSwingLine(&current_segment, &current_line, 0.0);
+            try decompressed_values.append(current_segment.start_value);
+            var current_timestamp: usize = current_segment.start_time + 1;
+            while (current_timestamp < current_segment.end_time) : (current_timestamp += 1) {
+                const y: f64 = f80Tof64(evaluate(&current_line, current_timestamp));
+                try decompressed_values.append(y);
+            }
+            try decompressed_values.append(current_segment.end_value);
+            first_timestamp = current_timestamp + 1;
+        } else {
+            try decompressed_values.append(current_segment.start_value);
+            first_timestamp += 1;
+        }
+    }
+}
+
+test "swing-filter single line compress and decompress" {
+    const alloc = testing.allocator;
```

Comment:
Most of the tests seems to follow the same structure, maybe move as much code as possible into helper function(s) to minimize duplicate code.

Reply:
This was fixed by removing tests that were covered by other bigger tests. Specifically, all the tests with fixed values were removed, and only the tests with random elements were left.